### PR TITLE
Garbage collector do not work on large tables

### LINKED ISSUE #153: Garbage collector do not work on large tables
I know it's probably something that is not supposed to happen, but if you are like me and you realize that you never cleaned the LoginSession table, running the Garbage collector is going to be a nightmare.

This is due to the fact it's using the orm to remove one by one every single record.

Even on a regular website, this is probably not great and using a couple of clean delete queries would be so much better.

I would like to propose to use direct DB queries instead if that something that would be considered since there is not much expectations about having orm extensions interfering with the login session anyway (side quest: don't use the orm at all for the login session queries).

Using the invalid all sessions task has similar issues

## PRs
- https://github.com/silverstripe/silverstripe-session-manager/pull/163
- https://github.com/silverstripe/developer-docs/pull/328

### PR DESCRIPTION
**Pull request by lekoala**
Fixes https://github.com/silverstripe/silverstripe-session-manager/issues/153

### CODE REVIEW

File: src/Services/GarbageCollectionService.php [hunk 1]
```diff
@@ -2,22 +2,50 @@
...
+    /**
+     * Limit the number of records collected per run. Defaults to 0 (no limit).
+     */
+    private static int $batch_remove_limit = 0;
+
     /**
      * Delete expired LoginSession and RememberLoginHash records
      */
-    public function collect(): void
+    public function collect(): array
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php]:
```suggestion
    public function collect(): void
```

File: src/Services/GarbageCollectionService.php [hunk 2]
```diff
@@ -2,22 +2,50 @@
...
-        $this->collectImplicitlyExpiredSessions();
-        $this->collectExpiredLoginHashes();
+        $expiredSessions = $this->collectExpiredSessions();
+        $implicitlyExpiredSessions = $this->collectImplicitlyExpiredSessions();
+        $loginHashes = $this->collectExpiredLoginHashes();
+        return [
+            'expiredSessions' => $expiredSessions,
+            'implicitlyExpiredSessions' => $implicitlyExpiredSessions,
+            'loginHashes' => $loginHashes,
+        ];
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php]:
```suggestion
        $this->collectExpiredSessions();
        $this->collectImplicitlyExpiredSessions();
        $this->collectExpiredLoginHashes();
```

File: src/Services/GarbageCollectionService.php [hunk 3]
```diff
@@ -30,27 +58,29 @@ private function collectExpiredSessions(): void
...
+        return $this->batchRemoveAll($sessions);
     }
 
     /**
      * Collect all persistent LoginSession records where the associated RememberLoginHash has expired
      */
     private function collectImplicitlyExpiredSessions(): void
     {
-        LoginSession::get()->filter([
+        $sessions = LoginSession::get()->filter([
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php:63]:
```suggestion
        LoginSession::get()->filter([
```

File: src/Services/GarbageCollectionService.php [hunk 4]
```diff
@@ -30,27 +58,29 @@ private function collectExpiredSessions(): void
...
             'Persistent' => 1,
             'LoginHash.ExpiryDate:LessThan' => date('Y-m-d H:i:s')
-        ])->removeAll();
+        ]);
+        return $this->batchRemoveAll($sessions);
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php]:
```suggestion
        $this->batchRemoveAll($sessions);
```

File: src/Services/GarbageCollectionService.php [hunk 5]
```diff
@@ -30,27 +58,29 @@ private function collectExpiredSessions(): void
             'LastAccessed:LessThan' => date('Y-m-d H:i:s', time() - $lifetime),
             'Persistent' => 0
         ]);
-        $sessions->removeAll();
+        return $this->batchRemoveAll($sessions);
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php]:
```suggestion
        $this->batchRemoveAll($sessions);
```

File: src/Services/GarbageCollectionService.php [hunk 6]
```diff
@@ -2,22 +2,50 @@
...
+        $i = 0;
+        $limit = self::config()->get('batch_remove_limit');
+        $limitedList = $limit > 0 ? $datalist->limit($limit) : $datalist;
+        DB::get_conn()->transactionStart();
+        foreach ($limitedList as $record) {
+            $record->delete();
+            $i++;
+        }
+        DB::get_conn()->transactionEnd();
+        return $i;
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php]:
```suggestion
    private function batchRemoveAll($datalist)
    {
        $limit = self::config()->get('batch_remove_limit');
        $limitedList = $limit > 0 ? $datalist->limit($limit) : $datalist;
        DB::get_conn()->transactionStart();
        foreach ($limitedList as $record) {
            $record->delete();
        }
        DB::get_conn()->transactionEnd();
```

File: src/Services/GarbageCollectionService.php [hunk 7]
```diff
@@ -30,27 +58,29 @@ private function collectExpiredSessions(): void
...
+        return $this->batchRemoveAll($sessions);
     }
 
     /**
      * Collect all RememberLoginHash records that have expired
      */
     private function collectExpiredLoginHashes(): void
     {
-        RememberLoginHash::get()->filter([
+        $hashes = RememberLoginHash::get()->filter([
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php:86]:
```suggestion
        RememberLoginHash::get()->filter([
```

File: src/Services/GarbageCollectionService.php [hunk 8]
```diff
@@ -30,27 +58,29 @@ private function collectExpiredSessions(): void
...
             'ExpiryDate:LessThan' => date('Y-m-d H:i:s')
-        ])->removeAll();
+        ]);
+        return $this->batchRemoveAll($hashes);
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php]:
```suggestion
```

File: src/Services/GarbageCollectionService.php [hunk 9]
```diff
@@ -2,22 +2,50 @@
...
 
 class GarbageCollectionService
 {
+    use Configurable;
     use Injectable;
 
+    /**
+     * Limit the number of records collected per run. Defaults to 0 (no limit).
+     */
+    private static int $batch_remove_limit = 0;
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php]:
```suggestion
     * Limit the number of records collected per run.
     */
    private static ?int $batch_remove_limit = null;
```
No need to mention the default - we can see it right there.

File: src/Services/GarbageCollectionService.php [hunk 10]
```diff
@@ -2,22 +2,50 @@
...
+            'implicitlyExpiredSessions' => $implicitlyExpiredSessions,
+            'loginHashes' => $loginHashes,
+        ];
+    }
+
+    private function batchRemoveAll($datalist): int
+    {
+        $i = 0;
+        $limit = self::config()->get('batch_remove_limit');
+        $limitedList = $limit > 0 ? $datalist->limit($limit) : $datalist;
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php]:
```suggestion
        $limit = self::config()->get('batch_remove_limit');
        $limitedList = $datalist->limit($limit);
```
With the default limit set to null, we can just trust the limit value outright.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Sorry, I didn't realise in your code sample in the issue that this would change the return type for some methods - we can't change return type as that will require anyone who has subclassed this class and overridden those methods to change their method signatures as well.

That said, given that the idea here is just to get the thing to _work_, I'd still be okay with the change from `removeAll()` to `batchRemoveAll()` without the API and output changes

**Comment by lekoala** [comment]
no worries, the change for the collect method was an added benefit. and instead of changing the return type, it's possible to store the result somewhere and add a getResults but that can be another PR. I still think it's valuable to allow to have some insights on what's being collected
For example in my cron tasks module, I store tasks results in the database. This would allow to know when you hit the limit, if you run this frequently enough, etc.

I've merged your changes and fixed some vars that went missing along the way

File: src/Services/GarbageCollectionService.php [hunk 11]
```diff
@@ -2,14 +2,23 @@
 
 namespace SilverStripe\SessionManager\Services;
 
+use SilverStripe\Core\Config\Configurable;
 use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\DB;
+use SilverStripe\ORM\FieldType\DBDatetime;
```
Comment (ChloeSartorelli) [inline src/Services/GarbageCollectionService.php:8]:
```suggestion
```
I was going to ignore this but you need to resolve merge conflicts anyway so may as well remove this unnecessary statement at the same time.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Looks good, can't see any issues with this now. Please resolve the conflict and I'll be happy to merge.

**Comment by GuySartorelli** [comment]
@lekoala Can you please resolve the merge conflict?

**Review by GuySartorelli — APPROVED** [review]
Thanks, looks good and works as expected.

**Pull request merged**